doc/neps/nep-0049.rst

```diff
@@ -0,0 +1,369 @@
+in size should not impact end-users.
+
+Detailed description
+--------------------
+
+High level design
+=================
+
+Users who wish to change the NumPy data memory management routines will use
+:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
+structure to hold pointers to functions used to manage the data memory. The
+calls are wrapped by internal routines to call :c:func:`PyTraceMalloc_Track`,
+:c:func:`PyTraceMalloc_Untrack`, and will use the
+:c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for
+auditing purposes. 
+
+Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
+that function may be called during the lifetime of an ``ndarray`` object, each
+``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
+time of its instantiation, and these will be used to reallocate or free the
+data memory of the instance. Internally NumPy may use ``memcpy` or ``memset``
+on the data ``ptr``.
+
+NumPy C-API functions
+=====================
+
+.. c:type:: PyDataMem_Handler
+
+    A struct to hold function pointers used to manipulate memory
+
+    .. code-block:: c
+
+        typedef struct {
+            char name[128];  /* multiple of 64 to keep the struct unaligned */
+            PyDataMem_AllocFunc *alloc;
+            PyDataMem_ZeroedAllocFunc *zeroed_alloc;
+            PyDataMem_FreeFunc *free;
+            PyDataMem_ReallocFunc *realloc;
+        } PyDataMem_Handler;
+
+    where the function's signatures are
+
+    .. code-block:: c
+
+        typedef void *(PyDataMem_AllocFunc)(size_t size);
+        typedef void *(PyDataMem_ZeroedAllocFunc)(size_t nelems, size_t elsize);
+        typedef void (PyDataMem_FreeFunc)(void *ptr, size_t size);
+        typedef void *(PyDataMem_ReallocFunc)(void *ptr, size_t size);
+        typedef void *(PyDataMem_CopyFunc)(void *dst, const void *src, size_t size);
+
+.. c:function:: const PyDataMem_Handler * PyDataMem_SetHandler(PyDataMem_Handler *handler)
+
+   Sets a new allocation policy. If the input value is NULL, will reset
+   the policy to the default. Returns the previous policy, NULL if the
+   previous policy was the default. We wrap the user-provided functions
+   so they will still call the python and numpy memory management callback
+   hooks.
+    
+.. c:function:: const char * PyDataMem_GetHandlerName(PyArrayObject *obj)
+
+   Return the const char name of the PyDataMem_Handler used by the
+   PyArrayObject. If NULL, return the name of the current global policy that
+   will be used to allocate data for the next PyArrayObject
+
+For an example of setting up and using the PyDataMem_Handler, see the test in
+:file:`numpy/core/tests/test_mem_policy.py`
+
+.. c:function:: void PyDataMem_EventHookFunc(void *inp, void *outp, size_t size, void *user_data);
+
+    This function will be called on NEW,FREE,RENEW calls in data memory
+    manipulation
```

Comment:
This is unclear to me...Do you mean `alloc`, `free`, and `realloc`?

Reply:
Or do you mean `PyDataMem_NEW` etc? If so, could you be explicit and add links to them?

Reply:
Removing this and the next function documentation, they are not part of the NEP.